File: leet-code/746.使用最小花费爬楼梯.js
/*
 * @lc app=leetcode.cn id=746 lang=javascript
 *
 * [746] 使用最小花费爬楼梯
 */

// @lc code=start
/**
 * @param {number[]} cost
 * @return {number}
 */
var minCostClimbingStairs = function (cost) {
  let dp = [cost[0], cost[1]]
  for (let i = 2; i < cost.length; i++) {
    const res1 = dp[0] + cost[i]
    const res2 = dp[1] + cost[i]
    dp[0] = dp[1]
    dp[1] = Math.min(res1, res2)
  }
  return Math.min(dp[0], dp[1])
};
// @lc code=end


// @after-stub-for-debug-begin
module.exports = minCostClimbingStairs;
// @after-stub-for-debug-end

File: leet-code-rs/src/solutions/746_使用最小花费爬楼梯.rs
/*
 * @lc app=leetcode.cn id=746 lang=rust
 *
 * [746] 使用最小花费爬楼梯
 */

// @lc code=start

#[allow(unused)]
impl Solution {
    pub fn min_cost_climbing_stairs(cost: Vec<i32>) -> i32 {
        if cost.len() <= 2 {
            return cost[0].min(cost[1]);
        }

        // dp[n]: 爬到下标 n 的最小花费
        // 转移：dp[n] = min(dp[n-1]+cost[n-1],dp[n-2]+cost[n-2])
        let mut dp = vec![0; cost.len() + 1];
        //0 或 1 开始
        // dp[0] = 0;
        // dp[1] = 0;

        for i in 2..=cost.len() {
            dp[i] = (dp[i - 1] + cost[i - 1]).min(dp[i - 2] + cost[i - 2]);
        }

        return dp[cost.len()];
    }
}
// @lc code=end

#[allow(unused)]
struct Solution;

#[cfg(test)]
mod test {
    #[allow(unused)]
    use super::*;

    #[test]
    fn test_min_cost_climbing_stairs() {
        //assert_eq!(Solution::min_cost_climbing_stairs(vec![]),[]);
        assert!(true)
    }
}
